### ISSUE TITLE
[Question] Different results using MultiInputPolicy and MlpPolicy with the same observation data

### ISSUE THREAD

**Issue by jhurlbut** [DLR-RM/stable-baselines3#444]
I am able to train my custom gym environment with good results using a MlpPolicy but when I change the policy to MultiInputPolicy and insert my vector observation array into a single element dictionary I get completely different results. From the tests it looks like I should get the same results as they are both using the FlattenExtractor. This is with a PPO policy.

Is there any guidance on what to check to see why the results are different? Thank you.

The changes made were:

Change the space to a dict:
```
self._observation_space = spaces.Box(low=-np.ones(self.num_obs) * np.inf, high=np.ones(self.num_obs) * np.inf, dtype=np.float32)
```
to
```
self._observation_space = spaces.Dict(
            spaces={
                "vec": spaces.Box(low=-np.ones(self.num_obs) * np.inf, high=np.ones(self.num_obs) * np.inf, dtype=np.float32)
            }
        )
```

Change the observation to a dict:
```
self._observation = np.zeros((self.num_envs, self.num_obs), dtype=np.float32)
```
to
```
self._observation = {"vec": np.zeros((self.num_envs, self.num_obs), dtype=np.float32)}
```        

Change the policy args from MlpPolicy to MultiInputPolicy:
```
 model = PPO('MlpPolicy', env, verbose=2, tensorboard_log=saver.data_dir)
```
to
```
model = PPO('MultiInputPolicy', env, verbose=2, tensorboard_log=saver.data_dir)
```

### Checklist

- [x] I have read the [documentation](https://stable-baselines3.readthedocs.io/en/master/) (**required**)
- [x] I have checked that there is no similar [issue](https://github.com/DLR-RM/stable-baselines3/issues) in the repo (**required**)
- [x] I have checked my env using the env checker (**required**)
- [x] I have provided a minimal working example to reproduce the bug (**required**)

**Comment by JadenTravnik** [comment]
Hey @jhurlbut, good question. 

I wrote out the below to verify that the difference between the initial policies would be the `features_extractor` module.
Hopefully this code represents your use-case, I ignored tensorboard as it shouldn't change the policy modules (but I haven't double checked that).

```
import gym
import numpy as np
import torch
from stable_baselines3 import PPO
from stable_baselines3.common.policies import MultiInputActorCriticPolicy, ActorCriticPolicy

class TestEnv(gym.Env):
    def __init__(self, is_dict=False, num_obs=10):
        self.action_space = gym.spaces.Discrete(1)
        obs_space = gym.spaces.Box(low=-np.ones(num_obs) * np.inf, high=np.ones(num_obs) * np.inf, dtype=np.float32)
        if not is_dict:
            self.observation_space = obs_space
        else:
            self.observation_space = gym.spaces.Dict({"vec_obs": obs_space})

    def reset(self):
        return self.observation_space.sample()

    def step(self, action):
        return self.observation_space.sample(), 0, False, {}

torch.manual_seed(0)
flat_model = PPO("MlpPolicy", TestEnv(is_dict=False), verbose=2)
torch.manual_seed(0)
dict_model = PPO("MultiInputPolicy", TestEnv(is_dict=True), verbose=2)

print("\n", flat_model.policy, dict_model.policy, sep="\n\n")

# asserts that the mlp_extractor weights are the same via the manual seed
assert (dict_model.policy.mlp_extractor.policy_net[0].weight == flat_model.policy.mlp_extractor.policy_net[0].weight).all()
assert (dict_model.policy.mlp_extractor.policy_net[2].weight == flat_model.policy.mlp_extractor.policy_net[2].weight).all()
```

Did you run both policies with multiple seeds? My first guess is that its the variability in the models training itself.
Could you try setting the torch seed when creating the policies to see if that has a difference on performance?

**Comment by araffin** [comment]
Hello,
this is weird as we have a consitency test for that type of example: https://github.com/DLR-RM/stable-baselines3/blob/df6f9de8f46509dad47e6d2e5620aa993b0fc883/tests/test_dict_env.py#L91

as @JadenTravnik wrote, please double check that you are doing the exact same experiment.

**Comment by jhurlbut** [comment]
Thanks for the replies and writing the test. Yes, I noticed the consistency test which is why I was confused by the issue. 

I have confirmed that the only code changes are the ones I posted. I can try running both models with multiple seeds. If the seeds are different I would think the performance would be similar but not exact. The issue is that the performance is really bad with the dict change as in the rewards never even reach 0. Here is a graph of the rewards with the MlpPolicy vs MultiInputPolicy

![MlpPolicy](https://user-images.githubusercontent.com/339204/119244672-a87fd280-bb27-11eb-9410-3a9bf1438559.png)
![MultiInputPolicy](https://user-images.githubusercontent.com/339204/119244675-ab7ac300-bb27-11eb-84ed-4f2419c73ceb.png)

I noticed that the multiinputpolicy and mlppolicy have different default learning rates. I'm not sure if that would cause the issue.

By wrappers are you referring to vector environments? I am running a vectorized environment. I can try changing the code to not be vectorized if you think that's a good thing to check. Thanks again for your guidance.

**Comment by araffin** [comment]
So I did additional testing (see below) and now I highly suspect that the issue may come from your env, probably some dict modified by reference...

>I noticed that the multiinputpolicy and mlppolicy have different default learning rates. I'm not sure if that would cause the issue.

What do you mean? They don't or it comes from your code then...

To test that, please use the wrapper below and please use the same seed for both runs:

```python
import gym
from stable_baselines3 import PPO

class DictWrapper(gym.Wrapper):
	def __init__(self, env):
		super(DictWrapper, self).__init__(env)
		self.observation_space = gym.spaces.Dict({
			"vec": self.env.observation_space
		})

	def reset(self):
		return {"vec": self.env.reset()}

	def step(self, action):
		obs, reward, done, infos = self.env.step(action)
		return {"vec": obs}, reward, done, infos
		


env = gym.make("CartPole-v1")


model = PPO("MlpPolicy", env, seed=8, verbose=1).learn(10000)


print()
print("=== Dict Env ===")
print()

env = DictWrapper(env)

model = PPO("MultiInputPolicy", env, seed=8, verbose=1).learn(10000)
```

**Comment by jhurlbut** [comment]
Using your DictWrapper class I get the same results with my custom environment with both the MlpPolicy and MultiInputPolicy so it fixes the issue.

If I wrap into a Monitor it threw an error and I had to comment out the wrapping into a DummyVecEnv. I think that is because the gym wrapper class casts my VecEnv to an Env and the isInstance test mistakenly thinks my VecEnv subclassed environment is an Env type.

Do you have any insight why using the DictWrapper works but changing the code inside my environment produces bad results?

Thank you

**Comment by araffin** [comment]
>my environment being a subclass of VecEnv 

Why are you doing that? Because you parallelize the processes using lower level language like c++ or using pytorch to batch computation?

Overall, I would recommend you to first check your env (that should derive from `gym.Env`) with the env checker (cf doc and issue template) before having something more complex and directly vectorizing the env from your side.

Giving multiple envs as input is ok but you must be careful with modifications by reference in python (take a closer look at `DummyVecEnv` to see how we deal with that).

**Comment by jhurlbut** [comment]
yes, the environment is using a c++ wrapper for some of the vectorizing processes. Thank you for the helpful advice.

**Comment by araffin** [comment]
>yes, the environment is using a c++ wrapper for some of the vectorizing processes. Thank you for the helpful advice.

the issue is now solved then ;)
